waiter/src/waiter/auth/jwt.clj

```diff
@@ -0,0 +1,332 @@
+                             :status status-unauthorized})))
+          (when (nil? exp)
+            (throw (ex-info "No expiry provided in the token payload"
+                            {:log-level :info
+                             :message "No expiry provided in the token payload"
+                             :status status-unauthorized})))
+          claims)))))
+
+(defn current-time-secs
+  "Returns the current time in seconds."
+  []
+  (-> (t/now) tc/to-long (/ 1000) long))
+
+(defn- access-token?
+  "Predicate to determine if an authorization header represents an access token."
+  [authorization]
+  (let [authorization (str authorization)]
+    (and (str/starts-with? authorization bearer-prefix)
+         (= 3 (count (str/split authorization #"\."))))))
+
+(defn extract-claims
+  "Returns either claims in the access token provided in the request, or
+   an exception that occurred while attempting to extract the claims."
+  [token-type issuer subject-key supported-algorithms key-id->jwk request]
+  (let [validation-timer (metrics/waiter-timer "core" "jwt" "validation")
+        timer-context (timers/start validation-timer)]
+    (try
+      (let [realm (utils/request->host request)
+            request-scheme (utils/request->scheme request)
+            bearer-entry (auth/select-auth-header request access-token?)
+            access-token (str/trim (subs bearer-entry (count bearer-prefix)))
+            claims (validate-access-token token-type issuer subject-key supported-algorithms key-id->jwk realm
+                                          request-scheme access-token)]
+        (timers/stop timer-context)
+        (counters/inc! (metrics/waiter-counter "core" "jwt" "validation" "success"))
+        claims)
+      (catch Throwable throwable
+        (timers/stop timer-context)
+        (counters/inc! (metrics/waiter-counter "core" "jwt" "validation" "failed"))
+        (log/info throwable "error in access token validation")
+        throwable))))
+
+(defn authenticate-request
+  "Performs authentication and then
+   - responds with an error response when authentication fails, or
+   - invokes the downstream request handler using the authenticated credentials in the request."
+  [request-handler token-type issuer subject-key supported-algorithms key-id->jwk password request]
+  (let [claims-or-throwable (extract-claims token-type issuer subject-key supported-algorithms key-id->jwk request)]
+    (if (instance? Throwable claims-or-throwable)
+      (if (-> claims-or-throwable ex-data :status (= status-unauthorized))
+        ;; allow downstream processing before deciding on authentication challenge in response
+        (ru/update-response
+          (request-handler request)
+          (fn [{:keys [status] :as response}]
+            (if (and (= status status-unauthorized) (utils/waiter-generated-response? response))
+              ;; add to challenge initiated by Waiter
```

Comment:
We're only sending challenges when the provided token is unauthorized, but we're falling through to another authentication method?  That feels odd.

I'm not sure what the desired behavior we're going for is.  As far as I see it:

1. JWT users
They want to speak JWT.  They indicate it by always providing the JWT.  If they provide an invalid JWT (say it's expired), they get challenged for a new, valid JWT.  They shouldn't get a challenged for another auth method (which seems to be what we're doing here.  Do you feel/think otherwise?)

2. Non-JWT users
They never provide JWTs, and they never get challenged for JWT. (This seems to be how things currently work in this PR, and this behavior is controlled elsewhere.)

Reply:
sradack wrote: 

> I'm confused the way we've set up the challenges however.
> 
> We're only sending challenges when the provided token is unauthorized, but we're falling through to another authentication method? That feels odd.
> I'm not sure what the desired behavior we're going for is. 
> As far as I see it:
> 
> JWT users
> They want to speak JWT. They indicate it by always providing the JWT. If they provide an invalid JWT (say it's expired), they get challenged for a new, valid JWT. They shouldn't get a challenged for another auth method (which seems to be what we're doing here. Do you feel/think otherwise?)
> 
> Non-JWT users
> They never provide JWTs, and they never get challenged for JWT. (This seems to be how things currently work in this PR, and this behavior is controlled elsewhere.)

Reply:
JWT authentication is treated as an optimization.
When a client provides multiple authorization headers, she is asking us to try one of many authentication schemes.
If JWT authentication fails, we can still try the configured authentication of the service (e.g. saml, kerberos, etc.).
If that authentication succeeds, then we can go ahead and process the response.

However, if that authentication fails too, none of the provided authentication schemes have worked.
We have a `401` unauthorized response and a challenge from the configured auth method.
We add in the challenge for JWT auth scheme too into the response.
The client can now decide how to tackle the challenge knowing that none of the provided authorization headers worked.